multi/mcxn94x-multi/cpu.c

```diff
@@ -0,0 +1,275 @@
+			the application reads from rx fifo. Also, fifo needs to be more
+			generic and at least have function to get its current length. */
+			lf_fifo_push(&common.rx_fifo, byte);
+		}
+	}
+
+	return (consumed > 0) ? 1 : -1;
+}
+
+
+static ssize_t fifoRead(unsigned char *buf, size_t size, unsigned int blocking)
+{
+	size_t bytes_read = 0;
+
+	if (buf == NULL || size == 0) {
+		return -EINVAL;
+	}
+
+	/* Synchronize multiple readers */
+	mutexLock(common.rx_lock);
+
+	if (blocking != 0) {
+		while (bytes_read < size) {
+			mutexLock(common.irq_lock);
+			while (lf_fifo_empty(&common.rx_fifo) != 0) {
+				condWait(common.rx_cond, common.irq_lock, 0);
+			}
+			mutexUnlock(common.irq_lock);
+
+			while (bytes_read < size) {
+				if (lf_fifo_pop(&common.rx_fifo, &buf[bytes_read]) != 0) {
+					bytes_read += 1;
+				}
+				else {
+					break;
+				}
+			}
+		}
+	}
+	else {
+		if (lf_fifo_empty(&common.rx_fifo) != 0) {
+			mutexUnlock(common.rx_lock);
+			return -EWOULDBLOCK;
+		}
+
+		while (bytes_read < size) {
+			if (lf_fifo_pop(&common.rx_fifo, &buf[bytes_read]) != 0) {
+				bytes_read += 1;
+			}
+			else {
+				break;
+			}
+		}
+	}
+
+	mutexUnlock(common.rx_lock);
+
+	return (ssize_t)bytes_read;
+}
+
+
+static ssize_t fifoWrite(const unsigned char *buf, size_t len, unsigned int blocking)
+{
+	size_t to_write;
+	size_t bytes_written = 0;
+
+	if (buf == NULL || len == 0) {
+		return -EINVAL;
+	}
+
+	/* Synchronize multiple writers */
+	mutexLock(common.tx_lock);
+
+	while (bytes_written < len) {
+		if (blocking == 0 && REG_IRQ(common.coreid_remote, reg_irq) != 0u) {
+			mutexUnlock(common.tx_lock);
+			return (bytes_written > 0) ? bytes_written : -EAGAIN;
+		}
+
+		while (blocking != 0 && REG_IRQ(common.coreid_remote, reg_irq) != 0u) {
+			; /* Wait until the other CPU reads IRQx register */
+		}
+
+		/* Message structure: MSB[ 6b reserved | 2b len(data) | 24b data ]LSB */
+		uint32_t value = 0;
+		to_write = min(len - (size_t)bytes_written, 3);
+
+		for (size_t i = 0; i < to_write; i++) {
+			value |= (uint32_t)buf[bytes_written] << (8 * i);
+			bytes_written += 1;
+		}
+
+		value |= ((uint32_t)to_write & 0x3) << 24;
+
+		REG_IRQ(common.coreid_remote, reg_irq_set) = value;
+	}
+
+	mutexUnlock(common.tx_lock);
+
+	return (ssize_t)bytes_written;
+}
+
+
+static int fifoPollStatus(void)
+{
+	int revents = 0;
+
+	if (lf_fifo_empty(&common.rx_fifo) == 0) {
+		revents |= POLLIN | POLLRDNORM;
+	}
+
+	if (REG_IRQ(common.coreid_remote, reg_irq) == 0u) {
+		revents |= POLLOUT | POLLWRNORM;
+	}
+
+	return revents;
+}
+
+
+static void cpu_handleMsg(msg_t *msg, msg_rid_t rid, unsigned int major, unsigned int minor)
+{
+	unsigned int blocking;
+
+	if (minor > 0) {
+		msg->o.err = -EINVAL;
+		msgRespond(msg->oid.port, msg, rid);
+		return;
+	}
+
+	switch (msg->type) {
+		case mtOpen:
+		case mtClose:
+			msg->o.err = 0;
+			break;
+
+		case mtWrite:
+			blocking = ((msg->i.io.mode & O_NONBLOCK) != 0u) ? 0u : 1u;
```

Comment:
Using bool for `blocking` would make code somewhat shorter in places

Reply:
Done